metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorker.java

```diff
@@ -175,171 +177,155 @@ private String convertRdfToStringForLogging(final RDF rdf) {
-        throw new DereferenceOrEnrichException(
-            "Exception occurred while trying to perform dereferencing.", e);
+      if (!foundOwnEntityIds.contains(resourceId)) {
+        result.addAll(dereferenceExternalEntity(resourceId));
       }
     }
-    return dereferenceInformation;
-  }
-
-  private static void logStringWithCounter(String field, int count) {
-    LOGGER.debug("== {}: {}", count, field);
-  }
 
-  private static void logInputValueWithCounter(InputValue inputValue, int count) {
-    LOGGER.debug("== {}: {} {} {} {}.", count, inputValue.getLanguage(),
-        inputValue.getOriginalField(), inputValue.getValue(), inputValue.getVocabularies());
+    // Done.
+    return result;
   }
 
-  private <T> void logExtractionResult(Collection<T> resultList, ObjIntConsumer<T> logger) {
-    int count = 0;
-    if (resultList != null) {
-      LOGGER.debug("Extracted following fields:");
-      for (T result : resultList) {
-        if (result == null) {
-          continue;
-        }
-        count++;
-        logger.accept(result, count);
-      }
-    }
-    if (count == 0) {
-      LOGGER.debug("No fields to process.");
+  private List<EnrichmentBaseWrapper> dereferenceOwnEntities(Set<String> resourceIds)
+          throws DereferenceOrEnrichException {
+    final EnrichmentResultList result;
+    try {
+      result = ExternalRequestUtil.retryableExternalRequest(() ->
+                      enrichmentClient.getById(resourceIds), mapWithRetrieableExceptions,
+              EXTERNAL_CALL_MAX_RETRIES, EXTERNAL_CALL_PERIOD_BETWEEN_RETRIES_IN_MILLIS);
+    } catch (Exception e) {
+      throw new DereferenceOrEnrichException(
+              "Exception occurred while trying to perform dereferencing.", e);
     }
+    return Optional.ofNullable(result).map(EnrichmentResultList::getEnrichmentBaseWrapperList)
+            .orElseGet(Collections::emptyList);
   }
 
-  private void logDereferencingOrEnrichmentResults(List<EnrichmentResultList> resultList) {
-    int count = 0;
-    if (resultList != null) {
-      LOGGER.debug("Following information found:");
-      for (EnrichmentResultList result : resultList) {
-        count += logDereferencingOrEnrichmentResult(result);
-      }
-    }
-    if (count == 0) {
-      LOGGER.debug("No information found. Nothing to merge.");
+  private List<EnrichmentBaseWrapper> dereferenceExternalEntity(String resourceId)
+          throws DereferenceOrEnrichException {
+    final EnrichmentResultList result;
+    try {
+      LOGGER.debug("== Processing {}", resourceId);
+      result = ExternalRequestUtil.retryableExternalRequest(() ->
+                      dereferenceClient.dereference(resourceId), mapWithRetrieableExceptions,
+              EXTERNAL_CALL_MAX_RETRIES, EXTERNAL_CALL_PERIOD_BETWEEN_RETRIES_IN_MILLIS);
+    } catch (BadRequest e) {
+      // We are forgiving for these errors
+      LOGGER.warn("ResourceId {}, failed", resourceId, e);
+      return Collections.emptyList();
```

Comment:
How about setting result to null and let the final return handle it?

Reply:
Sure, I guess that's fair.